crates/proto/src/udp/udp_stream.rs

```diff
@@ -239,58 +244,59 @@ impl<P: RuntimeProvider> Future for NextRandomUdpSocket<P> {
     /// polls until there is an available next random UDP port,
     /// if no port has been specified in bind_addr.
     fn poll(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output> {
-        if self.bind_address.port() == 0 {
-            // Per RFC 6056 Section 3.2:
-            //
-            // As mentioned in Section 2.1, the dynamic ports consist of the range
-            // 49152-65535.  However, ephemeral port selection algorithms should use
-            // the whole range 1024-65535.
-            let rand_port_range = Uniform::new_inclusive(1_024, u16::MAX);
-            let mut rand = rand::thread_rng();
-
-            for attempt in 0..10 {
-                let port = rand_port_range.sample(&mut rand);
-                let bind_addr = SocketAddr::new(self.bind_address.ip(), port);
-
-                // TODO: allow TTL to be adjusted...
-                // TODO: this immediate poll might be wrong in some cases...
-                match self
-                    .provider
-                    .bind_udp(bind_addr, self.name_server)
-                    .as_mut()
-                    .poll(cx)
-                {
+        let this = self.get_mut();
+        loop {
+            this.future = match this.future.take() {
+                Some(mut future) => match future.as_mut().poll(cx) {
                     Poll::Ready(Ok(socket)) => {
                         debug!("created socket successfully");
                         return Poll::Ready(Ok(socket));
                     }
                     Poll::Ready(Err(err)) => match err.kind() {
-                        io::ErrorKind::AddrInUse => {
-                            debug!("unable to bind port, attempt: {}: {}", attempt, err);
+                        io::ErrorKind::AddrInUse if this.attempted < ATTEMPT_RANDOM + 1 => {
+                            debug!("unable to bind port, attempt: {}: {err}", this.attempted);
+                            this.attempted += 1;
+                            None
                         }
                         _ => {
                             debug!("failed to bind port: {}", err);
                             return Poll::Ready(Err(err));
                         }
                     },
-                    Poll::Pending => debug!("unable to bind port, attempt: {}", attempt),
+                    Poll::Pending => {
+                        debug!("unable to bind port, attempt: {}", this.attempted);
+                        Some(future)
```

Comment:
Overall, this seems cleaner compared to #2462, but I have two (related) questions:

1. Couldn't this deadlock, at least until the future gets canceled?  The function returns on Poll::Ready(Ok) or a non-port-collision error, while the attempted counter is only incremented on a port-collision error.

2. Does this do the same thing as #2462, specifically as it relates to fixing the bug @mokeyish is seeing? @mokeyish's patch returns on Poll::Pending, whereas this doesn't appear to.

@mokeyish if you have a standalone minimal test case that reproduces the bug you're seeing it would be helpful, to me at least, to review these two proposed patches.

Reply:
1. Hmm, you're right, this should yield `Pending` if the inner future yields `Pending`. What happens after we return `Ready` doesn't really matter that much, though, because the future is only polled until `Ready` once.

2. I think the bug in #2462 is that it didn't preserve state in the `Pending` case -- for `Pending` in the randomized port case, we'd forget all state for the next invocation.

Reply:
@marcus0x62 My scenario is to set up a socks proxy. The previous version actually did not have await at all, causing the subsequent code to not be executed.

https://github.com/mokeyish/smartdns-rs/blob/188fca40f4d22506142c33617e05811e7bfd57db/src/proxy.rs#L27 

this line return a pendding state, which needs polling next time.

Reply:
Changed my last commit to yield `Pending` if the inner future yields `Pending`.